pubsub/server.go

```diff
@@ -0,0 +1,98 @@
+// Copyright (C) 2019-2022, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package pubsub
+
+import (
+	"net/http"
+	"sync"
+
+	"github.com/gorilla/websocket"
+
+	"go.uber.org/zap"
+
+	"github.com/ava-labs/avalanchego/utils/logging"
+)
+
+var upgrader = websocket.Upgrader{
+	ReadBufferSize:  readBufferSize,
+	WriteBufferSize: writeBufferSize,
+	CheckOrigin: func(*http.Request) bool {
+		return true
+	},
+}
+
+// Server maintains the set of active clients and sends messages to the clients.
+type Server struct {
+	log  logging.Logger
+	lock sync.RWMutex
+	// conns a set of all our connections
+	conns Connections
+	// Callback function when server receives a message
+	rCallback Callback
+	//  Supplemental info for the rCallback function
+	e []interface{}
+}
+
+// New returns a new Server instance with the logger set to [log]. The callback
+// function [f] is called by the server in response to messages if not nil.
+func New(log logging.Logger, r Callback, e []interface{}) *Server {
```

Comment:
One really nice feature would be to allow us to register handlers directly with this constructor. Then handlers we wouldn't need a root handler later that strips the first byte and routes elsewhere.

If you don't know what I mean, I can do this.